pkg/apis/apps/v1alpha1/types.go

```diff
@@ -0,0 +1,103 @@
+*/
+
+package v1alpha1
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+)
+
+// PetSet represents a set of pods with consistent identities.
+// Identities are defined as:
+//  - Network: A single stable DNS and hostname.
+//  - Storage: As many VolumeClaims as requested.
+// The PetSet guarantees that a given network identity will always
+// map to the same storage identity.
+type PetSet struct {
+	unversioned.TypeMeta `json:",inline"`
+	v1.ObjectMeta        `json:"metadata,omitempty"`
+
+	// Spec defines the desired identities of pets in this set.
+	Spec PetSetSpec `json:"spec"`
+
+	// Status is the current status of Pets in this PetSet. This data
+	// may be out of date by some window of time.
+	Status PetSetStatus `json:"status,omitempty"`
+}
+
+// A PetSetSpec is the specification of a PetSet.
+type PetSetSpec struct {
+	// Replicas is the desired number of replicas of the gieven Template.
+	// These are replicas in the sense that they are instantiations of the
+	// same Template, but individual replicas also have a consistent identity.
+	// TODO: Consider a rename of this field.
+	Replicas int `json:"replicas"`
+
+	// Selector is a label query over pods that should match the replica count.
+	// If empty, defaulted to labes on the pod template.
+	// More info: http://releases.k8s.io/HEAD/docs/user-guide/labels.md#label-selectors
+	Selector *unversioned.LabelSelector `json:"selector,omitempty"`
+
+	// Template is the object that describes the pod that will be created if
+	// insufficient replicas are detected. Each pod stamped out by the PetSet
+	// will fulfill this Template, but have a unique identity from the rest
+	// of the PetSet.
+	Template v1.PodTemplateSpec `json:"template"`
+
+	// VolumeClaims is a list of claims that pets are allowed to reference.
+	// The PetSet controller is responsible for mapping network identities to
+	// claims in a way that maintains the identity of a pet. Every claim in
+	// this list must have at least one matching (by name) volumeMount in one
+	// container in the template. A claim in this list takes precedence over
+	// any volumes in the template, with the same name.
+	// TODO: Define the behavior if a claim already exists with the same name.
+	VolumeClaims []v1.PersistentVolumeClaim `json:"volumeClaims,omitempty"`
+
+	// ServiceName is the name of the service that governs this PetSet.
+	// This service must exist before the PetSet.
+	// Pets get DNS/hostnames that follow the pattern:
+	// pet-specific-string.serviceName.default.svc.cluster.local
+	// where "pet-specific-string" is managed by the PetSet controller.
+	ServiceName string `json:"serviceName"`
+}
+
+// PetStatus represents the current state of a single pet in the set.
+type PetStatus struct {
+	// Name is the name of this particular pet.
+	Name string `json:"name,omitempty"`
+
+	// Volumes are the volumes associated with this pet.
+	Volumes []v1.Volume `json:"volumes,omitempty"`
```

Comment:
Would that not just be a label selector on the claims created?  I'd
imagined:

kubectl get petset
  petset-mine

kubectl get pods
  petset-mine-1
  petset-mine-2
  petset-mine-3

kubectl get pvc
  petset-mine-1-1
  petset-mine-1-2
  petset-mine-2-1
  petset-mine-2-2
  petset-mine-3-1
  petset-mine-3-2

with labels being applied to the pods and pvcs for grouping.  If you had
pet set size 100 (a reasonable high end for now) I'd rather have the volume
names be deterministic rather than try to show the list.  Controller only
has to fill in gaps, babysitter could use annotations on the petset for now
for locking if necessary (or on the service endpoints via discovery on the
serviceRef)


On Sat, Apr 16, 2016 at 12:45 AM, Prashanth B <notifications@github.com>
wrote:

> In pkg/apis/apps/v1alpha1/types.go
> <https://github.com/kubernetes/kubernetes/pull/24315#discussion_r59962449>
> :
>
> > +
> > +	// ServiceName is the name of the service that governs this PetSet.
> > +	// This service must exist before the PetSet.
> > +	// Pets get DNS/hostnames that follow the pattern:
> > +	// pet-specific-string.serviceName.default.svc.cluster.local
> > +	// where "pet-specific-string" is managed by the PetSet controller.
> > +	ServiceName string `json:"serviceName"`
> > +}
> > +
> > +// PetStatus represents the current state of a single pet in the set.
> > +type PetStatus struct {
> > +	// Name is the name of this particular pet.
> > +	Name string `json:"name,omitempty"`
> > +
> > +	// Volumes are the volumes associated with this pet.
> > +	Volumes []v1.Volume `json:"volumes,omitempty"`
>
> I figured giving more information about the volumes allocated
> (name+type+capacity) in describe would be better, since the petsest only
> includes a list of claims. I can make it a list of strings if you prefer?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24315/files/5fa8ac534e922d68815c7979c25304b714d6c0c5#r59962449>
>